### ISSUE TITLE
proposal: big.Decimal

### ISSUE THREAD

**Issue by kostya-sh** [golang/go#12127]
`math/big` will provide `big.Float` numeric type since go 1.5. However it uses binary floating point representation (`sign × mantissa × 2**exponent`).

In many areas, especially in finance, exact values need to be processed and the inputs are commonly decimal. Unfortunately, decimal values cannot be represented accurately using binary floating points.

Additionally many SQL databases support fixed size decimal types and currently Go doesn't provide data type out of the box to represent them.

Since Go can be used in many places where accurate decimal arithmetic is required it seems reasonable to support it by adding `big.Decimal` type to the standard Go library.

Many other languages and systems already provide similar functionality:

- `java.math.BigDecimal` in Java
- `decimal.Decimal` in Python
- `System.Decimal` in .NET

**Comment by cznic** [comment]
See also: http://godoc.org/github.com/shopspring/decimal

**Comment by rin01** [comment]
I have the same problem.

The reasons why decimal floating numbers are important can be found here:
http://speleotrove.com/decimal/decifaq.html

In my projects, I use the C library __decNumber__, written by Mike Cowlishaw.
http://speleotrove.com/decimal/

> The decNumber package, an implementation of the specifications in ANSI C, provides a reference implementation for both the arithmetic and the encodings.
> It includes both an arbitrary-precision implementation and a (much faster) decFloats implementation that uses the IEEE 754 decimal encodings directly to implement decSingle, decDouble, and decQuad datatypes.

To download the source files, go to "International Components for Unicode (ICU)" and click "decNumber package".

The decQuad type is 128 bits long, and stores a decimal floating point number.
decDouble type is 64 bits long, and decSingle type is 32 bits long.

Personnally, I just put the .c and .h files in my project tree, and call the library functions using cgo.
I just use the __decQuad__ type, which has the largest precision and stores numbers in 128 bits.
By experience, I really think it is an excellent library.

Example of use:
http://speleotrove.com/decimal/dnusers.html#example7

It would be great if a Go package wrapping this C library could be integrated in Go standard library, before being replaced by a pure Go solution with the same API.
In this case, it should perhaps be called "decnumber".

Note that this library is already integrated in __gcc compiler__.
But the type name have changed, e.g. decQuad became __Decimal128 in gcc.
https://gcc.gnu.org/onlinedocs/gcc/Decimal-Float.html

> As an extension, GNU C supports decimal floating types as defined in the N1312 draft of ISO/IEC WDTR24732.
> Support for decimal floating types in GCC will evolve as the draft technical report changes.
> Calling conventions for any target might also change. Not all targets support decimal floating types.
> 
> The decimal floating types are _Decimal32, _Decimal64, and _Decimal128.

**Comment by griesemer** [comment]
math/big is all about arbitrary precision arithmetic, and in the case of Float, the precision can be set arbitrarily (within memory limits, essentially).

I'd be ok with a big.Decimal type; as long as it's also arbitrary precision. Fixed-size data types such as decQuad, decDouble belong into a separate package in my mind.

In fact, the Float implementation already uses some (minimal) decimal arithmetic for Float->string conversion; having an actual Decimal type would make that code unnecessary (in favor of using a fully-fledged Decimal implementation).

That said, I would like to see more "meat" around this proposal. With the Float implementation, the fixed-width IEEE floating-point spec provided reasonable guidance. What are the parameters for a Decimal implementation? Is it at Fixed-point implementation (arbitrary size number with fixed decimal point), or decimal floating-point representation (like Float, with mantissa and exponent, but the mantissa is always decimal)? Are the other options? What are the pros and cons? What are other packages doing? What is required for say business/banking applications?

**Comment by ianlancetaylor** [comment]
Just a note: any fixed-size Decimal implementation should consider the IEEE 754-2008 specs for decimal floating point, and the corresponding C extensions for decimal floating point (http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1312.pdf).

**Comment by rin01** [comment]
@kostya-sh: do you really need __arbitrary precision__ for financial computation ?
In my mind, they are best done by using base 10 floating point __fixed-size__ data types, like a decimal128 (or decQuad).

I say that because I think the API of big.Float (and big,Decimal will certainly be the same) is quite complex.

As an example, for addition:
> func (z *Float) Add(x, y *Float) *Float
> Add sets z to the rounded sum x+y and returns z. If z's precision is 0, it is changed to the larger of x's or y's precision before the operation. Rounding is performed according to z's precision and rounding mode; and z's accuracy reports the result error relative to the exact (not rounded) result. 

You must prepare the result variable with proper precision and rounding mode, and think about it for each arithmetic operation.

With a fixed-size floating point API, and decimal128 data type, all you do would be:

> c = context.Add(a, b)

where context contains the rounding mode, and stores computation error of it occurs.
__Also, numbers are just values, not pointers.__
It would be as simple as working with float64.

The API of big.Float introduces a complexity that is not needed for most financial computation.
We just want that Add, Subtract and Mult don't round the result for numbers in usual range, but store them with the max precision allowed by the fixed-size data type. e.g. 34 significant digits for decimal128.
And if we want to round or truncate the result, we can just call the proper function.

### Usual range for monetary values

Assets under management for a large multinational insurance company is:

$265,507,000,000

This figure is the sum of amounts that can have 8 digits after decimal point, because this precision is needed when working with foreign currency conversion.

This means that for a large company, a decimal data type must cope with figures like:

__100,000,000,000.00000000__, that is, 20 significant digits.

A decimal128 fixed-size data type provides about 34 significant digits, which can easily store these figures.
We can even store USA national debt of $18,000,000,000,000 in it with full precision.

When doing financial computation, intermediate calculation can require some more digits, but __34 significant digits is really enough for most real cases financial computation__.

That's why I think that big.Decimal for most financial applications is overkill, and a fixed-size decimal128 base-10 floating type is more than "good enough".

### PRO and CONS

__big.Decimal__ arbitrary-precision 

CON: if like big.Float, the API is more complex than needed
CON: variables are pointers
CON: precision and rounding mode of the result must be think about for each operation
CON: is there a lot of financial applications that requires a precision of more than 34 significant digits ?
PRO: no limit to the precision

__decnumber.Decimal128__ fixed precision

PRO: API is simple
PRO: variable are values
PRO: no need to think about precision, as the result is always the max precision of decimal128
PRO: decimal128 can store 34 significant digits, which can accomodate the range of values used in most financial applications
PRO: no need to allocate memory to store the result, all numbers fits in 128 bits. (use of sync.Pool and small internal buffer can alleviate this problem, though)

That being said, any means of working with base-10 floating point numbers, be it arbitrary-precision or fixed-size, will be welcomed by all those that work with monetary values ;-)

**Comment by kostya-sh** [comment]
@rin01, I agree that for the most (if not all) financial computations decimal128 is enough. This is approach is used by .NET. `System.Decimal` is decimal128 though I am not sure how closely it follows IEEE 754-2008.

However I think that arbitrary precision `big.Decimal` type is a better fit for the inclusion in the Go standard library as it supports more use cases than decimal128. For example:

- Non-financial use-cases. E.g. `big.Float` to string conversion.
- Representing arbitrary precision numbers that are supported in other systems such as SQL databases. For example [PostgreSQL decimal](http://www.postgresql.org/docs/9.1/static/datatype-numeric.html) can store up to 131072 digits before the decimal point and up to 16383 digits after the decimal point.

`big.Decimal` API should probably be very similar to `big.Float` API for consistency. This means that `big.Decimal` will be mutable and its methods will be defined using pointers. To address some of your concerns:

- **Mutable vs immutable**. I believe this is a matter of user preference. Some users find surprising that calling `a.add(b, context)` in Java doesn't modify `a`. From the performance point of view mutable values leave more space for optimization, at least as far as memory usage is concerned.
- **Values vs Pointers**. It would be good to be able to work with large number of decimals without any GC overhead. While it won't be possible to avoid it in all cases (because of arbitrary precision) I believe that for the most common usecase when precision is small GC can be avoided completely by using `uint64` instead of `big.Int` as `big.Decimal` internal representation. Go compiler should take care of allocating `big.Decimal` instances on the stack when possible using escape analysis.

**Comment by kostya-sh** [comment]
The following proposal documents can be used to understand what other packages are doing:
- [JSR 13](https://www.jcp.org/en/jsr/detail?id=13) for Java. Interestingly the original author seems to be Mike Cowlishaw.
- [PEP 0327](https://www.python.org/dev/peps/pep-0327/) for Python

Both Java and Python implementations are based on a decimal arithmetic model, as defined by the relevant standards: IEEE 854 , ANSI X3-274 and the proposed [revision](http://grouper.ieee.org/groups/754/revision.html) of IEEE 754 .

**Comment by rin01** [comment]
In Go, numeric data types have both a built-in fixed-size implementation, and a package for arbitrary-precision.

Ultimately, Go should perhaps have both kinds.
A default fixed-size decimal128 implementation for usual cases, and a big.Decimal for more specialized cases.

base | built-in fixed-size   | arbitrary-precision package
---- | ------------ | -------------------
2    | int          | big.Int
2    | float64      | big.Float
10   | __decimal128__   | __big.Decimal__

**Comment by rin01** [comment]
If anyone is interested in seing what a fixed-size decimal package looks like, I have quickly written this Go wrapper around the decNumber package of Mike Cowllishaw, but just the __decQuad__ data type.

__decnum.DecQuad is a 128 bits fixed-size base-10 floating point__

I have not tested it, it is just a quick (but not dirty) work.

https://github.com/rin01/decnum

https://godoc.org/github.com/rin01/decnum

**Comment by kostya-sh** [comment]
@rin01, your decnum package looks very interesting. In fact I think that a package that supports both fixed size types (`decimal.Single`, `decimal.Double`, `decimal.Quad`) and an arbitrary precision type (`decimal.Number`) for decimal numbers that is modeled after [decnumber](http://speleotrove.com/decimal/decnumber.html) library could be a very good addition to _golang.org/x_ packages. Such package will probably will have to be implemented in pure Go. I don't think that it fits `math/big` package though as the APIs are quite different. Maybe you can create another proposal?

If possible I would like to keep this proposal limited to the new `math/big` `Decimal` type with a set of methods similar to `Float` type. The purpose of `big.Decimal` is to provide a standard and consistent API to work with arbitrary precision decimal numbers in Go that will work for 99% of users. And users requiring more control or better performance or different API could look at other packages similar to your `decnum` proposal.

I will update the proposal with more details as requested by @griesemer soon.

**Comment by kostya-sh** [comment]
I have updated the issue to include more details about the proposed implementation.

**Comment by rin01** [comment]
@kostya-sh: I agree with you. This proposal should be limited to math/big Decimal type with similar API to Float type. This would be great and will make the __arbitrary-precision__ package math/big complete in a consistent manner.

__Fixed-size__ base-10 floating point like decimal128 is another topic and should belong to another package and issue.
Just for info, I am working on my decnum package to make it fully tested, debugged and usable for next week.
This way, even if a pure Go implementation of decimal128 (aka decnum.Quad) is desirable, it won't be an urgent topic any more.

**Comment by mpvl** [comment]
As an FYI, some of the issues that are aimed to be solved by this proposal, like rounding modes and precision (trailing zeros), I plan to address in the number formatting package in the text repo. In this model, aspects like rounding and precision are considered a property of the visualization, not the underlying number. This allows precision in computation to be higher than that of visualization. The number rendering would also support the big variants.

This is not to say that an addition of big.Dec (naming to be consistent) wouldn't be useful; I think it would be, even though it could be easily simulated with big.Int.  Mostly, it would define a standard for decimal numbers so that it can be supported out of the box in other repos (like text). Note that many major frameworks, including .Net, OS X Foundation, Java etc. all have this functionality in their core library, which indicates it probably belongs there.

One could try to make the rounding definitions more general than just math/big.  One could define a math/round package (as in .Net) or, IMO better, define the rounding modes and behaviors in math and add a func Round(float64, Rounding) (res float64, result bool) to complement Floor and Trunc. In math/big, the various operations would use these rounding types. The method Dec(round math.Rounding) (d *Dec, exact bool) could then be defined on Rat to convert it to Dec. There is no efficient easy way to do that now that I can think of.

Note that these rounding parameters don't cover the whole spectrum of rounding algorithms. For currencies and in finance one sometimes does not round to a power of 10. I think it will be sufficient to simulate it, though.  For example, rounding for Swiss accounting could be achieved by rounding HalfEven and then completing the rounding "manually" to select 0 or 5 for the last digit.

**Comment by kostya-sh** [comment]
@mpvl, thank you for your feedback. Please find my comments below.

Precision and rounding mode should be available to operations on big.Decimal because of the following reasons:
- very often it is practical to limit precision during calculations. E.g. result of multiplication of 3 numbers with 5 digits after decimal point can have 5^3=625 digits after decimal point. If precision is unbounded operations will become slow. It is less error prone to set precision once instead of rounding results after each operation.
- to implement decimal32, decimal64, decimal128 fixed point IEEE 754 types and operations

Number of trailing zeroes is determined by the scale, not precision. The precision is the upper limit on the number of significant digits in a result. E.g. for 1.2000, scale is 4 and precision could be any number >=2. 

I prefer Decimal name rather than Dec but I will leave the decision to golang team.

I agree that `math` is a better place for `RoundingMode` (and probably `Accuracy`) than `math/big` as it is applicable not only to the arbitrary precision numbers but also to float32 and float64 and the proposed Round function might be useful addition to `math` package. I am not sure if this still can be changed. @griesemer ?

Methods to convert between big.Decimal and float32, float64, big.Int, big.Float and big.Rat will be part of this proposal.

Various accounting rounding rules are not described in IEEE 754 standard and don't have to be supported by big.Decimal. It should be easy to implement them if necessary though. For example last 2 digits required for Swiss rounding (as described at (http://blog.syspro.com/2013/09/12/rounding-accounting) could be calculated by setting scale to 4 and getting reminder of division by 0.01.

**Comment by griesemer** [comment]
@kostya-sh Regarding your question: Both the math and math/big package are subject to the Go compatibility rules - thus we can add stuff but in general we cannot change or remove existing API. One could define Accuracy and RoundingMode in math and redefine the same constants in math/big. But let's not get bogged down by implementation details for now.

I'm worried that this proposal's scope is growing rather than becoming more focused. There's two questions to be answered:

1) Is this a proposal for a commercially viable decimal number/fixed-point package?
2) Is this a proposal for an arbitrary precision decimal package?

If the goal is to make Go a viable language for business/commercial applications (which I think would be a great goal), than a first step might be an easy-to-use, standards-compliant implementation of a fixed-point decimal package (from what I understand so far, 128bit seems plenty). Such a package doesn't need to be under math or math/big. It could be the base of a larger library of business-related packages.

If the goal is an arbitrary precision decimal type, I think it would simply be an extension of package big (as in big.Decimal). 

I think the proposal should focus on one of the two, I'm not convinced that they need to be combined. My vote would be for making this about 1): A proposal for commercially viable decimal number/fixed-point package. This is the more interesting one, with potential for large impact if done well.

Instead of focusing on a concrete implementation, the proposal should define exactly what would be needed for commercial applications. Business/accounting software experts, anyone?

**Comment by kostya-sh** [comment]
@griesemer as far as I know Accuracy and RoundingMode are new types in Go 1.5. The only reason I asked the question is that now (before 1.5 final release) is the only time they still can be moved without breaking compatibility. I don't have a strong opinion on the location of these types though and if you think that they are defined in the right package let them be.

I agree with you that the scope of this proposal should be limited. I saw a few questions on StackOverflow ([1](http://stackoverflow.com/questions/31946344/why-does-go-treat-a-postgresql-numeric-decimal-columns-as-uint8), [2](http://stackoverflow.com/questions/31503742/how-do-i-format-a-currency-with-commas-and-2-decimal-places), [3](http://stackoverflow.com/questions/19238175/what-is-the-best-way-to-convert-a-currency-from-float-to-integer-in-go), [4](http://stackoverflow.com/questions/20596428/how-to-represent-currency-in-go), [5](http://stackoverflow.com/questions/31778173/what-type-to-use-to-correctly-handle-division-of-odd-numbers-of-cents-or-small), [6](http://stackoverflow.com/questions/26008004/go-type-for-purchasing-financial-calculations)) that would be very easy to answer if Go had big.Decimal type. That was my original motivation for creating this proposal.

I am by no means an expert but I have some experience working with financial (though not accounting) applications in Java. It provides arbitrary precision decimal class (java.math.BigDecimal) in the core library and it is recommended to use this class for monetary calculations. It also integrates nicely with decimal database types via JDBC API. These are the reasons I believe that `big.Decimal` should be good enough for the most commercial (and not only) applications. The main complaints about BigDecimal in Java are:
- Verbose API. Go (similar to Java) doesn't support operator overloading though, so it is unlikely that the Go API can be made significantly less verbose.
- Poor performance (comparing to native floating point types). Still It is still reasonably fast and applications that need to be as fast as possible tend to use double or long types to represent amounts (e.g. http://vanillajava.blogspot.com/2011/08/double-your-money-again.html)

I would prefer to keep this proposal limited to big.Decimal and create another proposal for "a commercially viable decimal number/fixed-point package". It probably should be discussed on the mailing list as proposals on github issue tracker are visible only to the limited number of interested developers.

**Comment by griesemer** [comment]
@kostya-sh The time for 1.5 API changes has long passed, especially with the final release any day now.

It's ok to make this about math/big.Decimal. In that case, I think the questions about Accuracy, RoundingMode are already answered. They are already where you'd want them.

So the rest of the proposal could then talk about internal representation, supported operations, etc.

**Comment by kostya-sh** [comment]
I have updated the original proposal with more details as requested.

**Comment by adg** [comment]
Why not implement this as an external package? Then we can see how it looks and how generally useful it is before talking about including it in the standard library.

**Comment by mpvl** [comment]
@adg Having a big.Decimal in core allows other core packages, like database/sql to refer to it. You won't have this benefit with an external package. In general standardizing on this type seems quite useful.

**Comment by griesemer** [comment]
@mpvl I'd be ok with a Decimal type in math/big, but I agree with @adg that prototyping it outside makes a lot of sense. I think the basic proposal (functionality, etc.) seems reasonable, but the devil is in the details, and some things may only become clear when an implementation is attempted.

An external implementation may be (slightly) less efficient since it cannot use say big.nat - but it's probably marginal. Prototyping it externally also removes any time pressure from getting it done for a specific release.

**Comment by kostya-sh** [comment]
@griesemer , I agree with your points. I've created https://github.com/kostya-sh/go-decimal-proposal and will work on it my spare time. Once it is more or less ready I will update this issue

**Comment by rin01** [comment]
@kostya-sh , have you looked at this package http://godoc.org/speter.net/go/exp/math/dec/inf ?
Working on it is perhaps less work that starting from scratch...

https://groups.google.com/forum/#!topic/golang-nuts/Zu_gSp56N4Y

**Comment by rsc** [comment]
See [my reply](https://github.com/golang/go/issues/12332#issuecomment-152540264) on #12332. If a group of developers want to work together in golang.org/x/exp/decimal on a type decimal.Big, that seems like a good next step.

**Comment by kostya-sh** [comment]
I have started to work on https://github.com/kostya-sh/go-decimal-proposal though by now I should admit that I do not have enough time and interest to finish it in reasonable timeframe by myself. Having said that I would be glad to contribute to golang.org/x/exp/decimal if it is created.

**Comment by faide** [comment]
Hi all,

I know I arrive a bit late to the party but I'd like to add a few comments.

I am a seasoned Python dev working mainly with Python in the financial / accounting  realm.
I am also using Go and in my search for a good python decimal.Decimal [1] equivalent in Go I found an interesting implementation [2]. The guys are using it in production to handle their online sales.

I am using it myself for some price related computations. I did not know about decnum [3] but will definitely look into it, thanks !

I would be interested to know what people interested in the decimal proposition think of shopspring/decimal

Cheers,
Florent

1: https://docs.python.org/3.5/library/decimal.html
2: https://github.com/shopspring/decimal
3: https://github.com/rin01/decnum

**Comment by kostya-sh** [comment]
I had a quick look at shopspring/decimal code. It is a useful library and the fact that it is used in production proves that it provides enough functionality for common use cases. However for edge cases the results can be surprising:

- `decimal.New(1, -100000000).Cmp(decimal.New(-1, 0))` will take forever to finish and will use a lot of memory
- `decimal.New(1, 0).Div(decimal.New(1, 100))` is 0. DivisionPrecision =16 and is global.

Some important functionality is missing as well:

- no support for different rounding modes for Div
- no support for fixed precision decimals
- no support for string conversion to exponential notation
- no support for Infinity or NaN (this can be considered a feature though)

**Comment by EricLagergren** [comment]
Hey all. I've been working on my own decimal library for a little while and it's mostly completed.

It's a "big" decimal type -- not decimal32, 64, 128, etc.

My current goals are:

1. Finish comprehensive test suite (and iron out any bugs that pop up)
2. Add proper support for scientific notation strings
3. Figure out how I want to support NaN/Inf (currently panics but since it's a "big" decimal that's hard to do) in a nice way
4. Minor optimizations (some parts are a bit naive)

Please feel free to rip it apart (or contribute) if you'd like. I'm a bit biased, but I believe it's a more proper library than some of the others out there. (Don't get me wrong -- I'm not knocking theirs, but the ones I've seen have to convert decimals to strings and parse the strings to do division or use cgo.)

https://github.com/EricLagergren/decimal

- Eric

**Comment by kostya-sh** [comment]
@EricLagergren, this looks great. If you need data for the comprehensive test suite feel free to reuse some of the tests that I've created for https://github.com/kostya-sh/go-decimal-proposal (based on the test cases from http://speleotrove.com/decimal).

In particular you can take a look at https://github.com/kostya-sh/go-decimal-proposal/tree/master/dectest - a small program that can generate go files with test tables from http://speleotrove.com/decimal/dectest.pdf. You will have to modify it to support more operations though.

A few comments:
- You use int64 for scale, I don't think this will work on 32-bit architectures.
- It would be good to add support for `fmt.Printf` (for `%s` and `%d`), fmt.Scanf
- I think some methods are unnecessary, like Fib or Binomial. I would use a big.Int-based library for them.
- MarshalJSON and UnmarshalJSON are not needed. MarshalText/UnmarshalText are enough

**Comment by EricLagergren** [comment]
@kostya-sh Thank you!

> You use int64 for scale, I don't think this will work on 32-bit architectures.

 Could you elaborate why? If there are issues (I haven't tested on 32-bit platforms yet) using `int` would probably work fine.

> I think some methods are unnecessary, like Fib or Binomial. I would use a big.Int-based library for them.

I added Fib (and, obviously, FizzBuzz) just for the fun of it :smile: 

> MarshalJSON and UnmarshalJSON are not needed. MarshalText/UnmarshalText are enough

I  still need to nail down this part of the library, although it should be easy enough.

**Comment by kostya-sh** [comment]
> Could you elaborate why? If there are issues (I haven't tested on 32-bit platforms yet) using int would probably work fine.

Just to clarify, it is possible to use int64 type to represent scale even on 32-bit platforms, however big.Int with the corresponding number of decimal digits simply won't fit in available 2Gb of memory. Actually I doubt it will even fit in memory on 64-bit platforms ;) So using int64 for scale and precision is a bit pointless.

int32 (not int) should be enough. Also it makes Decimal smaller (by 8 bytes - 4 for prec, 4 for scale) and some operations are easier to implement correctly. E.g. in Cmp you have ` hi.scale - lo.scale` (https://github.com/EricLagergren/decimal/blob/master/decimal.go#L493). However you do not check for overflow. If scale is int32 then checking for overflow is trivial as you can do all arithmetic operations using int64 and compare the final result with max.MaxInt32

**Comment by EricLagergren** [comment]
> So using int64 for scale and precision is a bit pointless.

Pfft, 32-bit plebs ;)

Also, FWIW while positive scales *do*, negative scales don't actually "use" as much memory as they'd appear to. `New(1234, -1e9)` doesn't suck up 1e9 bytes of memory -- all that's used will be what Decimal needs (so, like 40 bytes or so). The exponent is implied. For a positive scale, yes, it would need 1e9 bytes of memory. :cry: 

> E.g. in Cmp you have hi.scale - lo.scale [...] However you do not check for overflow.

That line is assuming `max(a, b) - min(a, b)` which won't over or underflow.

I do think using int32 would be a good idea... and I think I could cut out some extra code since I would have a few more assumptions I could make regarding over and underflows.

It'd be great to have another set of eyes on the project -- would you happen have a free moment to go over it? That'd be excellent.

**Comment by kostya-sh** [comment]
> It'd be great to have another set of eyes on the project -- would you happen have a free moment to go over it? That'd be excellent.

Sure, I will be glad to help. I will create a few issues with my feedback in https://github.com/EricLagergren/decimal

**Comment by EricLagergren** [comment]
@kostya-sh Awesome! Thank you -- I really appreciate it.

**Comment by EricLagergren** [comment]
If one was to begin working on an exp/decimal, where would one begin?

**Comment by adg** [comment]
@EricLagergren if you want to work in golang.org/x/exp, then you need to find someone to review your code (and ideally that is someone who works on other parts of Go, already). But there's no reason to let that hold you back; you can just work on the code in a repository hosted anywhere.

**Comment by EricLagergren** [comment]
@adg I have github.com/ericlagergren/decimal

I've submitted code to the stdlib but I didn't know if creating something in /x/exp required anything other than a CL.

**Comment by adg** [comment]
@EricLagergren yep, it's just a matter of sending a CL. But there needs to be at least one other person willing to work with you on it (do code reviews, etc).

**Comment by EricLagergren** [comment]
@adg I threw up a CL if you're at all interested :-)

**Comment by adg** [comment]
@EricLagergren thanks for doing that, but I don't have the bandwidth to review it.

**Comment by EricLagergren** [comment]
@adg no worries, just in case you were curious is all. Cheers.

**Comment by adg** [comment]
I took a look at the CL, and replied there. An initial CL should be more substantial: it should include the basic API and documentation, at least.

**Comment by EricLagergren** [comment]
@adg The API has been outlined with comments for each routine. Apologies for the lacking CL. I've a ton more code to throw up whenever it's needed.

**Comment by adg** [comment]
@EricLagergren thanks. That looks like a fine start. However, as I said earlier, you really need another interested contributor to work with you on this. Otherwise there's no difference to just using the one you already have on your GitHub account.

**Comment by EricLagergren** [comment]
@adg yep, I've been meaning to make a post on golang-nuts or golang-dev when I have a free minute. Thanks for giving the CL a look over, I appreciate it.

**Comment by ruimarinho** [comment]
Is there any de facto library outside the main repository you can recommend @rsc?

**Comment by EricLagergren** [comment]
@ruimarinho there's https://github.com/ericlagergren/decimal, although I'm not trying to give out russ' blessings for him :smile:

**Comment by rsc** [comment]
I have not used any decimal implementations myself. In addition to @EricLagergren's, the next-to-last comment currently on #12332 reports success using https://github.com/shopspring/decimal. As usual, any code you rely upon in important, production systems should first undergo thorough testing to ensure that it meets your needs.

**Comment by anshupitlia** [comment]
which one did you use @ruimarinho ?

**Comment by brobits** [comment]
I am still running into the need for larger precision exact decimal type.  Google, while searching use cases, please realize Go is being used for many cryptocurrency projects.  Half these projects appear to internally use float64s, half use shopspring.Decimal.  Please support something similar to C++'s `decimal128` at least, the need is real.

**Comment by ericlagergren** [comment]
@brobits Here are some alternatives I feel comfortable recommending (in a slightly biased order :-)

- github.com/ericlagergren/decimal
- github.com/cockroachdb/apd

Most of the other decimal libraries don't have robust tests or skim on some fundamental aspects of the GDA spec. I share your sympathies.

**Comment by ianlancetaylor** [comment]
@brobits This issue is closed.  Commenting here won't do anything.

Also, this issue is about an unlimited precision fixed point decimal type.  It's not about `decimal128`.  That is issue #12332, which is also closed.

The place to discuss this topic is the golang-nut mailing list.
